spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -286,7 +288,15 @@ trait SnapshotManagement { self: DeltaLog =>
       // Here we validate that we are able to create a valid LogSegment by just using commit deltas
       // and without considering minor-compacted deltas. We want to fail early if log is messed up
       // i.e. some commit deltas are missing (although compacted-deltas are present).
-      validateDeltaVersions(deltasAfterCheckpoint, newCheckpointVersion, versionToLoad)
+      // We should not do this validation when we want to update postCommitSnapshot. In that
+      // specific flow, we just list from the committed version and reuse existing pre-commit
+      // logsegment together with listing result to create the new snapshot (Refer to
+      // [[SnapshotManagement.getUpdatedLogSegment]]). That codepath explicitly needs to disable
+      // validation as the validation is expected to fail in that as we don't have all delta file
+      // info when minor compactions are used in pre-commit snapshot.
+      if (validateLogSegmentWithoutCompactedDeltas) {
+        validateDeltaVersions(deltasAfterCheckpoint, newCheckpointVersion, versionToLoad)
```

Comment:
I don't understand why the check needs to be(come) conditional? [validateDeltaVersions](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala#L191-L217) doesn't seem to care whether every version has a commit .json file available? It seems to just verify that the list of commits and/or minor compactions covers all needed versions (in order) after the checkpoint. The existing log segment should trivially pass that check (because we checked when constructing it), and it seems like we still want to validate the new files in relation to old ones?

Could you elaborate on how/why "validation is expected to fail" in that case?

Reply:
Update: `deltasAfterCheckpoint` _only_ contains deltas (see L255-256 above, which separates checkpoints, deltas, and minor compactions). Normal listing returns both deltas and minor compactions for the same commits, but log segments only include one or the other. Thus, the check would wrongly report gaps in the version sequence if the existing log segment included any minor compactions.

Downside is, by disabling the check we risk missing any problems in the newly listed deltas after the log segment. But I guess it's only an extra sanity check, so maybe not worth tearing up the code to keep track of the separation point between existing vs. new files?

Reply:
Yes - keeping the check for getUpdatedLogSegment will need us to maintain the separation point - which doesn't seem to be worth the complexity.